### PR TITLE
[#1] Add neofs backend

### LINKED ISSUE #1: Add NeoFS backend
Define NeoFS backend in internal/backend catalogue. It should implement

- [ ] Location
- [ ] Hasher
- [ ] Test
- [ ] Remove
- [ ] Close
- [ ] Save
- [ ] Load
- [ ] Stat
- [ ] List
- [ ] IsNotExist
- [ ] Delete

### PR DESCRIPTION
**Pull request by KirillovDenis**
I'm not sure if we need use pool or plain client.
closes #1 

Signed-off-by: Denis Kirillov <denis@nspcc.ru>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -19,19 +19,24 @@ require (
 	github.com/minio/minio-go/v7 v7.0.14
 	github.com/minio/sha256-simd v1.0.0
 	github.com/ncw/swift/v2 v2.0.0
+	github.com/nspcc-dev/neo-go v0.96.1
```
Comment (alexvanin) [inline go.mod]: go for v0.97.3

File: internal/backend/neofs/config.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+type Config struct {
+	Endpoint  string
+	Container string
+
+	Wallet            string        `option:"wallet" help:"path to the wallet"`
+	Address           string        `option:"address" help:"address of account (can be empty)"`
+	Password          string        `option:"password" help:"password to decrypt wallet"`
+	Timeout           time.Duration `option:"timeout" help:"timeout to connect and request (default 10s)"`
+	RebalanceInterval time.Duration `option:"rebalance" help:"interval between checking node healthy (default 15s)"`
+	Policy            string        `option:"policy" help:"placement policy (default 'REP 3')"`
```
Comment (alexvanin) [inline internal/backend/neofs/config.go]: Backups are quite important thing and placement policy really matter there. Maybe restic should not create containers for backups at all. Let user control it. I would rather return error if expected container is not available.

File: internal/backend/neofs/config.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	"time"
+
+	"github.com/restic/restic/internal/errors"
+	"github.com/restic/restic/internal/options"
+)
+
+// Config contains all configuration necessary to connect to neofs.
+type Config struct {
+	Endpoint  string
+	Container string
```
Comment (alexvanin) [inline internal/backend/neofs/config.go:15]: While having container nice name is good feature, I would expect an option to specify container ID value as well.

Maybe we can parse both `neofs:grpcs://s01.neofs.devenv:8080/container-name` and `neofs:grpcs://s01.neofs.devenv:8080/container-id` ?

File: internal/backend/neofs/config.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	RebalanceInterval time.Duration `option:"rebalance" help:"interval between checking node healthy (default 15s)"`
+	Policy            string        `option:"policy" help:"placement policy (default 'REP 3')"`
+
+	SessionExpiration uint64
+}
+
+// NewConfig returns a new Config with the default values filled in.
+func NewConfig() Config {
+	return Config{
+		//Connections: 5,
```
Comment (alexvanin) [inline internal/backend/neofs/config.go]: debug?

File: internal/backend/neofs/neofs.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+	p, err := createPool(ctx, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	containerID, err := getContainerID(ctx, p, cfg.Container, cfg.Policy)
+	if err != nil {
+		return nil, err
+	}
+	fmt.Println("container repo", containerID.String())
```
Comment (alexvanin) [inline internal/backend/neofs/neofs.go]: debug? By the way don't see any problems with using `debug.Logs` in the code.

File: internal/backend/neofs/config.go (hunk 2 above)
Comment (alexvanin) [reply internal/backend/neofs/config.go]: /cc @realloc

**Comment by alexvanin** [comment]
> I'm not sure if we need use pool or plain client.

Extra endpoints can be specified as the parameter string, so I would go for pool even though it is pool of one client right now.

**Comment by alexvanin** [comment]
There were a few NeoFS errors, such as
```
Save(<data/09875faee5>) returned error, retrying after 310.599456ms: closing the stream failed: rpc error: code = Unknown desc = could not receive response: could not close stream and receive response: (*object.putStreamResponser) could not receive response: (*response.ClientMessageStreamer) could not close stream and
 receive response: (*putsvc.streamer) could not object put stream: (*putsvc.Streamer) could not close object target: could not close target: could not close next target: incomplete object PUT by placement
```

Maybe they are caused by docker networking,  because automatic retries were always solving the problem.

File: internal/backend/neofs/config.go (hunk 3 above)
Comment (KirillovDenis) [reply internal/backend/neofs/config.go:15]: Fixed

**Comment by KirillovDenis** [comment]
> There were a few NeoFS errors, such as
> 
> ```
> Save(<data/09875faee5>) returned error, retrying after 310.599456ms: closing the stream failed: rpc error: code = Unknown desc = could not receive response: could not close stream and receive response: (*object.putStreamResponser) could not receive response: (*response.ClientMessageStreamer) could not close stream and
>  receive response: (*putsvc.streamer) could not object put stream: (*putsvc.Streamer) could not close object target: could not close target: could not close next target: incomplete object PUT by placement
> ```
> 
> Maybe they are caused by docker networking, because automatic retries were always solving the problem.

I've got the same a couple times. But it disappeared when I replaced plain client with the pool so I thought it was related to correct using session token or something like that. But now it seems I was wrong.

**Review by alexvanin — APPROVED** [review]

**Pull request merged**